src/engine/sync/abletonlink.cpp

```diff
@@ -0,0 +1,232 @@
+#include "engine/sync/abletonlink.h"
+
+#include <QtDebug>
+#include <cmath>
+
+#include "control/controlobject.h"
+#include "control/controlpushbutton.h"
+#include "engine/sync/enginesync.h"
+#include "moc_abletonlink.cpp"
+#include "preferences/usersettings.h"
+#include "util/logger.h"
+namespace {
+const mixxx::Logger kLogger("AbletonLink");
+constexpr mixxx::Bpm kDefaultBpm(9999.9);
+} // namespace
+
+AbletonLink::AbletonLink(const QString& group, EngineSync* pEngineSync)
+        : m_group(group),
+          m_pEngineSync(pEngineSync),
+          m_mode(SyncMode::None),
+          m_oldTempo(kDefaultBpm),
+          m_audioBufferTimeMicros(0),
+          m_absTimeWhenPrevOutputBufferReachesDac(0),
+          m_sampleTimeAtStartCallback(0) {
+    m_pLink = std::make_unique<ableton::Link>(120.);
+    m_timeAtStartCallback = m_pLink->clock().micros();
+
+    m_pLinkButton = std::make_unique<ControlPushButton>(ConfigKey(group, "sync_enabled"));
+    m_pLinkButton->setButtonMode(ControlPushButton::TOGGLE);
+    m_pLinkButton->setStates(2);
+
+    connect(m_pLinkButton.get(),
+            &ControlObject::valueChanged,
+            this,
+            &AbletonLink::slotControlSyncEnabled,
+            Qt::DirectConnection);
+
+    m_pNumLinkPeers = std::make_unique<ControlObject>(ConfigKey(m_group, "num_peers"));
+    m_pNumLinkPeers->setReadOnly();
+    m_pNumLinkPeers->forceSet(0);
+
+    // The callback is invoked on a Link - managed thread.
+    // The callback is the only entity, which access m_pNumLinkPeers
+    m_pLink->setNumPeersCallback([this](std::size_t numPeers) {
+        m_pNumLinkPeers->forceSet(numPeers);
+    });
+
+    m_pLink->enable(false);
+
+    // Start/Stop sync makes not much sense for a DJ set
+    m_pLink->enableStartStopSync(false);
+
+    nonAudioThreadDebugOutput();
+    audioThreadDebugOutput();
+
+    initTestTimer(1000, true);
+}
+
+AbletonLink::~AbletonLink() {
+    m_pLink.reset();
+}
+
+void AbletonLink::slotControlSyncEnabled(double value) {
+    qDebug() << "AbletonLink::slotControlSyncEnabled" << value;
```

Comment:
Sorry if I'm missing something, but isn't that exactly the point of it? The debug messages shouldn't show up in production logs and thus not add much overhead in that case.

Reply:
yes, debug logs are general a good thing. But this line in particular does make much sense because it seems like a leftover from printf-debugging:
1. new code made for properly debugging user issues should leverage logging categories. Logging categories are overkill for simple printf-debugging but are really useful in logs IMO.
2. This line is just logging whether a slot is called. this doesn't seem particularly useful in the general case where we want debug messages to show up in user logs so we can trace an issue.